Update lisk-dpos to Typescript - Closes #4669

ManuGowda opened the pull request:
### What was the problem?
`lisk-dpos` needed to be migrated to typescript #4669 
### How did I solve it?
Update lisk-dpos to Typescript 
### How to manually test it?
- Build and run lisk-sdk test application
- Run lisk-dpos tests
### Review checklist

- [ ] The PR resolves #4669 
- [ ] All new code is covered with unit tests
- [ ] Relevant integration / functional tests are added
- [ ] Commit messages follow the [commit guidelines](CONTRIBUTING.md#git-commit-messages)
- [ ] Documentation has been added/updated

shuse2 commented on elements/lisk-dpos/test/fixtures/constants.ts:
```diff
@@ -0,0 +1,20 @@
+ * Removal or modification of this copyright notice is prohibited.
+ */
+
+'use strict';
```
no need to `use strict` in TS file

shuse2 commented on elements/lisk-dpos/tsconfig.json:
```diff
@@ -4,7 +4,7 @@
+		"types": ["node", "jest", "jest-extended"]
 	},
-	"include": ["src/**/*"]
+	"include": ["src/**/*", "types/**/*"]
```
no need to add "types/**/**` here

pablitovicente commented on elements/lisk-dpos/src/constants.ts:
```diff
@@ -14,8 +14,4 @@
 
 'use strict';
```
I think TS does not need `use strict` ?

pablitovicente commented on elements/lisk-dpos/src/delegates_list.ts:
```diff
@@ -14,16 +14,37 @@
 	const seedSource = round.toString();
 	const delegateList = [...list];
+	// tslint:disable-next-line:no-let
 	let currentSeed = hash(seedSource, 'utf8');
```
why not const here and tslint disabled?

pablitovicente commented on elements/lisk-dpos/src/delegates_list.ts:
```diff
@@ -36,21 +57,40 @@ const shuffleDelegateListForRound = (round, list) => {
+	public async getForgerPublicKeysForRound(
+		round: number,
+		delegateListRoundOffset?: number,
+		tx?: StorageTransaction,
```
if this `tx` refers to a database transaction it would be better to find a better name to avoid the conflict with the term `transaction` on our domain

pablitovicente replied:
oh well maybe this was already like this so it's ok if too much effort to rename now.

pablitovicente commented on elements/lisk-dpos/src/delegates_list.ts:
```diff
@@ -63,7 +103,9 @@ class DelegatesList {
 
-	async getDelegatePublicKeysSortedByVoteWeight(tx) {
+	public async getDelegatePublicKeysSortedByVoteWeight(
+		tx?: StorageTransaction,
```
if this `tx` refers to a database transaction it would be better to find a better name to avoid the conflict with the term `transaction` on our domain

pablitovicente replied:
oh well maybe this was already like this so it's ok if too much effort to rename now.

ManuGowda commented:
@shu @pablo I am aware that since 2.1 typescript version use strict is not required, compiler emits them. I removed them initially and added back as we were keeping in other elements library.  We should remove it from all element libraries, for now I'll remove from mine and I'll create an issue to remove from other as well. Danke !!

pablitovicente commented on elements/lisk-dpos/src/delegates_info.ts:
```diff
@@ -164,26 +281,36 @@ class DelegatesInfo {
+					acc: AccountFees,
+					{ delegateAccount, earnings: { fee, reward } }: UniqueForgerInfo,
+				) => {
+					(delegateAccount as Account).votedDelegatesPublicKeys.forEach(
```
Else TSLint fails

ManuGowda commented on elements/lisk-dpos/src/delegates_info.ts:
```diff
@@ -136,54 +234,67 @@ class DelegatesInfo {
+				) => {
+					(delegateAccount as Account).votedDelegatesPublicKeys.forEach(
+						publicKey =>
+							(acc[publicKey] = acc[publicKey]
```
I think it was assigning, no?
```
if (acc[publicKey]) {
						acc[publicKey] = acc[publicKey].plus(fee.plus(reward));
					} else {
						acc[publicKey] = fee.plus(reward);
					}
```

mitsuaki-u replied:
Ah you're right sorry the extra parenthesis threw me off, I think the outer parenthesis not needed.
